File: study/ts/script/data/GrainCarUnitData.ts
import { ConfigNameConst } from "../const/ConfigNameConst";
import { G_ConfigLoader, G_ServerTime, G_UserData } from "../init";
import GrainCarConfigHelper from "../scene/view/grainCar/GrainCarConfigHelper";
import { assert } from "../utils/GlobleFunc";
import { BaseData } from "./BaseData";


export interface GrainCarUnitData {
    getId():number
    setId(data:number):void
    getLevel():number
    setLevel(data:number):void
    getExp():number
    setExp(data:number):void
    getStamina():number
    setStamina(data:number):void
    getStart_time():number
    setStart_time(data:number):void
    getEnd_time():number
    setEnd_time(data:number):void
    getAll_view():number
    setAll_view(data:number):void
    getGuild_id():number
    setGuild_id(data:number):void
    getGuild_name():number
    setGuild_name(data:number):void
    getDonate_users():number
    setDonate_users(data:number):void
    getConfig():any
    setConfig(data:object):void
    getDestory_guild_name():object
    setDestory_guild_name(data:object):void
    getGrainRoads():any
    setGrainRoads(data:object):void
    getRoute():any
    setRoute(data:object):void
    getRoad_id():number
    setRoad_id(data:number):void
    getRoute_point():any
    setRoute_point(data:object):void
    getRoute_passed():object
    setRoute_passed(data:object):void
    getMine_id():number
    setMine_id(data:number):void
    
}

var schema = {};
schema['id'] = [
    'number',
    0
];
schema['level'] = [
    'number',
    0
];
schema['exp'] = [
    'number',
    0
];
schema['stamina'] = [
    'number',
    0
];
schema['start_time'] = [
    'number',
    0
];
schema['end_time'] = [
    'number',
    0
];
schema['all_view'] = [
    'number',
    0
];
schema['guild_id'] = [
    'number',
    0
];
schema['guild_name'] = [
    'string',
    0
];
schema['donate_users'] = [
    'number',
    0
];
schema['config'] = [
    'object',
    {}
];
schema['destory_guild_name'] = [
    'object',
    {}
];
schema['grainRoads'] = [
    'object',
    {}
];
schema['route'] = [
    'object',
    {}
];
schema['road_id'] = [
    'number',
    0
];
schema['route_point'] = [
    'object',
    {}
];
schema['route_passed'] = [
    'object',
    {}
];
schema['mine_id'] = [
    'number',
    0
];

export  class  GrainCarUnitData extends BaseData {

    public static schema = schema;

    constructor(properties?) {
        super(properties);
        this["grainRoads"] = {};
    }
    clear() {
    }
    reset() {
    }
    initData(msg) {
        this.setProperties(msg);
        this.setLevel(msg.id);
       
        var config = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR).get(msg.id);
        assert(config, 'graincar can\'t find base_id = ' + (msg.id));
        this.setConfig(config);
    }
    updateData(data) {
        this.setProperties(data);
        this.setLevel(data.id);
        var config = G_ConfigLoader.getConfig(ConfigNameConst.GRAINCAR).get(data.id);
        assert(config, 'graincar can\'t find base_id = ' + (data.id));
        this.setConfig(config);
    }
    insertData(data) {
    }
    deleteData(data) {
    }
    createRouteWithRoadId(roadId) {
        if (roadId && roadId > 0) {
            this.setRoad_id(roadId);
            var routeList = GrainCarConfigHelper.getRouteWithId(roadId);
            this.setRoute_point(routeList);
        }
    }
    setGrainRoads(road) {
        this["grainRoads"] = road
        this._initPassedRouteList();
    }
    getGrainRoads(){
        return this["grainRoads"];
    }
    isInMine(mineId) {
        var roads = this.getGrainRoads();
        if (!roads.length||roads.length < 1) {
            return null;
        }
        var curRoad = roads[0];
        var curTime = G_ServerTime.getTime();
        if (curRoad.getMine_id() == mineId && curTime >= curRoad.getEnter_time() - 1 && curTime < curRoad.getLeave_time()) {
            return true;
        }
        return false;
    }
    getCurPit() {
        var roads = this.getGrainRoads();
        if (!roads.length||roads.length < 1) {
            return null;
        }
        var curRoad = roads[0];
        var minePit1 = curRoad.getMine_id();
        return minePit1;
    }
    getCurCarPos():Array<any> {
        var roads = this.getGrainRoads();
        if (!roads.length||roads.length < 1) {
            return [null,null,null];
        }
        var curRoad = roads[0];
        var minePit1 = curRoad.getMine_id();
        var minePit2 = curRoad.getNext_mine_id();
        var percent = 0;
        var interval = this.getConfig().moving;
        var curMSTime = G_ServerTime.getMSTime();
        if (curMSTime < curRoad.getLeave_time() * 1000) {
            percent = 0;
        } else if (curMSTime > curRoad.getLeave_time() * 1000 + interval * 1000) {
            percent = 1;
        } else {
            percent = (curMSTime - curRoad.getLeave_time() * 1000) / (interval * 1000);
        }
        return [
            minePit1,
            minePit2,
            percent
        ];
    }
    getRoutePercent() {
        var curTime = G_ServerTime.getTime();
        return (curTime - this.getStart_time()) / (this.getEnd_time() - this.getStart_time());
    }
    getStartMine() {
        var roads = this.getRoute();
        if (roads.length < 1) {
            return null;
        }
        var curRoad = roads[0];
        var minePit = curRoad.getMine_id();
        return minePit;
    }
    isStop() {
        var roads = this.getGrainRoads();
        if (!roads.length||roads.length < 1) {
            return null;
        }
        var curRoad = roads[0];
        return G_ServerTime.getTime() < curRoad.getLeave_time() || curRoad.getLeave_time() == 0;
    }
    isOnWay() {
        var [_, _, percent] = this.getCurCarPos();
        return percent > 0 && percent < 1;
    }
    isReachTerminal() {
        var roads = this.getGrainRoads();
        if ((!roads.length||roads.length < 1) && this.getStamina() > 0) {
            return true;
        }
        var curRoad = roads[0];
        if (curRoad.getNext_mine_id() == 0 && this.getStamina() > 0) {
            return true;
        }
        return false;
    }
    getLeaveTime() {
        var roads = this.getGrainRoads();
        var curRoad = roads[0];
        var leaveTime = curRoad.getLeave_time();
        return leaveTime;
    }
    isFriendCar() {
        var guildId = G_UserData.getGuild().getMyGuildId();
        return this.getGuild_id() == guildId;
    }
    hasLaunched() {
        return this.getStart_time() > 0;
    }
    isActive() {
        var curTime = G_ServerTime.getTime();
        return curTime > this.getStart_time() && curTime < this.getEnd_time() && this.getStamina() > 0;
    }
    isDead() {
        return this.getStamina() <= 0;
    }
    hasComplete() {
        var curTime = G_ServerTime.getTime();
        return curTime >= this.getEnd_time() && this.getStamina() > 0;
    }
    getNextRouteList() {
        var routeList = this.getRoute_point();
        var i = 0;
        var curPit = this.getCurPit();
        while (i < routeList.length) {
            var pit = routeList[i];
            if (curPit != pit) {
                routeList.splice(i,1);
                i = i - 1;
            } else {
                break;
            }
            i = i + 1;
        }
        return routeList;
    }
    _initPassedRouteList() {
        var routePassed = [];
        var routeList = this.getWholeRoute();
        var i = 0;
        var curPit = this.getCurPit();
        while (i < routeList.length) {
            var pit = routeList[i];
            routePassed.push(pit);
            if (curPit == pit) {
                break;
            }
            i = i + 1;
        }
        this.setRoute_passed(routePassed);
    }
    getWholeRoute() {
        if (this.getRoad_id() > 0) {
            return GrainCarConfigHelper.getRouteWithId(this.getRoad_id());
        } else {
            var routeList = [];
            var roadList = this.getRoute();
            for (var i = 0; i < roadList.length; i++) {
                var road = roadList[i];
                var minePit = road.getMine_id();
                routeList.push(minePit);
            }
            return routeList;
        }
    }
}

File: study/lua/src/app/data/GrainCarUnitData.lua
-- Description: 粮车数据
-- Company: yoka
-- Author: chenzhongjie
-- Date: 2019-09-06

local BaseData = require("app.data.BaseData")
local GrainCarUnitData = class("GrainCarUnitData", BaseData)
local GrainCarConfigHelper = require("app.scene.view.grainCar.GrainCarConfigHelper") 

local schema = {}
schema["id"]					= {"number", 0} --粮车configId 就是等级 ————贝贝
schema["level"]					= {"number", 0} --粮车configId 就是等级 ————贝贝
schema["exp"]					= {"number", 0} --经验
schema["stamina"] 				= {"number", 0} --耐久度
schema["start_time"]			= {"number", 0} --开始移动的时间
schema["end_time"]				= {"number", 0} --位移结束的时间（开始存的是未被摧毁最终结束时间）
schema["all_view"]				= {"number", 0} --军团所有人可见 (0:管理 1:所有人)
schema["guild_id"]				= {"number", 0} --军团id
schema["guild_name"]		    = {"string", 0} --军团名字
schema["donate_users"]			= {"number", 0} --军团捐献的人次
schema["config"]            	= {"table", {}} --配置表信息
schema["destory_guild_name"]    = {"table", {}} --伤害我方粮车的军团名列表
schema["grainRoads"]            = {"table", {}} --路线信息
schema["route"]                 = {"table", {}} --完整路线
schema["road_id"]               = {"number", 0} --路线id
schema["route_point"]           = {"table", {}} --后续路径点（客户端算）
schema["route_passed"]          = {"table", {}} --已经经过的路径点
schema["mine_id"]               = {"number", 0} --死亡点


GrainCarUnitData.schema = schema


function GrainCarUnitData:ctor(properties)
	GrainCarUnitData.super.ctor(self, properties)
end

function GrainCarUnitData:clear()
end

function GrainCarUnitData:reset()
end

function GrainCarUnitData:initData(msg)
    self:setProperties(msg)
    self:setLevel(msg.id)
    local config = require("app.config.graincar").get(msg.id)
    assert(config, "graincar can't find base_id = " .. tostring(msg.id))
    self:setConfig(config)
end

function GrainCarUnitData:updateData(data)
    self:setProperties(data)
    self:setLevel(data.id)
    local config = require("app.config.graincar").get(data.id)
    assert(config, "graincar can't find base_id = " .. tostring(data.id))
    self:setConfig(config)
end

function GrainCarUnitData:insertData(data)
end

function GrainCarUnitData:deleteData(data)
end

function GrainCarUnitData:createRouteWithRoadId(roadId)
    if roadId and roadId > 0 then
        self:setRoad_id(roadId)
        local routeList = GrainCarConfigHelper.getRouteWithId(roadId)
        self:setRoute_point(routeList)
    end
end

function GrainCarUnitData:setGrainRoads(road)
    self.grainRoads_ = road
	self:_initPassedRouteList()
end

--是否在某个mine_id里停留
function GrainCarUnitData:isInMine(mineId)
    local roads = self:getGrainRoads()
    if #roads < 1 then
        return false
    end
    local curRoad = roads[1]
    local curTime = G_ServerTime:getTime()
    if curRoad:getMine_id() == mineId and 
        curTime >= curRoad:getEnter_time() - 1 and
        curTime < curRoad:getLeave_time() then
        return true
    end
    return false
end

--获取当前矿id（路上也算）
function GrainCarUnitData:getCurPit()
    local roads = self:getGrainRoads()
    if #roads < 1 then
        return nil
    end
    local curRoad = roads[1]
    local minePit1 = curRoad:getMine_id()
    return minePit1
end

--获取路线百分比
--return minePit1, minePit2, percent
function GrainCarUnitData:getCurCarPos()
    local roads = self:getGrainRoads()
    if #roads < 1 then
        return nil
    end
    local curRoad = roads[1]
    local minePit1 = curRoad:getMine_id()
    local minePit2 = curRoad:getNext_mine_id()
    local percent = 0
    local interval = self:getConfig().moving --GrainCarConfigHelper.getGrainCarMoveTime()
    local curMSTime = G_ServerTime:getMSTime()
    if curMSTime < curRoad:getLeave_time() * 1000 then
        percent = 0 
    elseif curMSTime > curRoad:getLeave_time() * 1000 + interval * 1000 then
        percent = 1
    else
        percent = (curMSTime - curRoad:getLeave_time() * 1000) / (interval * 1000)
    end
    return minePit1, minePit2, percent
end

--获取全程路线百分比(活着有效)
function GrainCarUnitData:getRoutePercent()
    local curTime = G_ServerTime:getTime()
    return (curTime - self:getStart_time()) / (self:getEnd_time() - self:getStart_time())
end

--获取起点矿
function GrainCarUnitData:getStartMine()
    local roads = self:getRoute()
    if #roads < 1 then
        return nil
    end
    local curRoad = roads[1]
    local minePit = curRoad:getMine_id()
    return minePit
end

--是否停留状态
function GrainCarUnitData:isStop()
    local roads = self:getGrainRoads()
    if #roads < 1 then
        return nil
    end
    local curRoad = roads[1]
    return G_ServerTime:getTime() < curRoad:getLeave_time() or curRoad:getLeave_time() == 0
end

--是否在路上
function GrainCarUnitData:isOnWay()
    local _, _, percent = self:getCurCarPos()
    return percent > 0 and percent < 1
end

--是否已经到终点
function GrainCarUnitData:isReachTerminal()
    local roads = self:getGrainRoads()
    if #roads < 1 and self:getStamina() > 0 then
        return true
    end
    local curRoad = roads[1]
    if curRoad:getNext_mine_id() == 0 and self:getStamina() > 0 then
        return true
    end
    return false
    -- local _, pit2, percent = self:getCurCarPos()
    -- local mineData = G_UserData:getMineCraftData():getMineDataById(pit2):getConfigData()
    -- return mineData.pit_type == 2 and percent == 1
end

--获取出站时间
function GrainCarUnitData:getLeaveTime()
    local roads = self:getGrainRoads()
    local curRoad = roads[1]
    local leaveTime = curRoad:getLeave_time()
    return leaveTime
end

--是否是友军粮车
function GrainCarUnitData:isFriendCar()
    local guildId = G_UserData:getGuild():getMyGuildId()
    return self:getGuild_id() == guildId
end

--是否已经发车
function GrainCarUnitData:hasLaunched()
    return self:getStart_time() > 0
end

--是否在活跃中 （没死且没跑完全程）
function GrainCarUnitData:isActive()
    local curTime = G_ServerTime:getTime()
    return curTime > self:getStart_time() and 
            curTime < self:getEnd_time() and 
            self:getStamina() > 0
end

--是否已经死亡
function GrainCarUnitData:isDead()
    return self:getStamina() <= 0
end

--是否已经跑完全程
function GrainCarUnitData:hasComplete()
    local curTime = G_ServerTime:getTime()
    return curTime >= self:getEnd_time() and self:getStamina() > 0
end

--获取接下来的路径点 包括当前点
--{1,2,3,4,5,6,7,8,9} 当前为5，则返回{5,6,7,8,9}
function GrainCarUnitData:getNextRouteList()
    local routeList = self:getRoute_point()
    local i = 1
    local curPit = self:getCurPit()
	while i < #routeList do
        local pit = routeList[i]
		if curPit ~= pit then
			table.remove(routeList, i)
            i = i - 1
        else
            break
		end
		i = i + 1
    end
    return routeList
end

--初始化已经经过的路径点，包括当前点
--{1,2,3,4,5,6,7,8,9} 当前为5，则返回{1,2,3,4,5}
function GrainCarUnitData:_initPassedRouteList()
    local routePassed = {}
    local routeList = self:getWholeRoute()
    local i = 1
    local curPit = self:getCurPit()
	while i < #routeList do
        local pit = routeList[i]
        table.insert(routePassed, pit)
		if curPit == pit then
            break
		end
		i = i + 1
    end
    self:setRoute_passed(routePassed)
end

--获取完整路线
function GrainCarUnitData:getWholeRoute()
    if self:getRoad_id() > 0 then
        return GrainCarConfigHelper.getRouteWithId(self:getRoad_id())
    else
        local routeList = {}
        local roadList = self:getRoute()
        for i = 1, #roadList do
            local road = roadList[i]
            local minePit = road:getMine_id()
            table.insert(routeList, minePit)
        end
        return routeList
    end
end


return GrainCarUnitData